profiler/upload_test.go

```diff
@@ -133,11 +134,89 @@ func BenchmarkDoRequest(b *testing.B) {
 		host:     "my-host",
 		profiles: []*profile{&prof},
 	}
-	p := unstartedProfiler()
+	p, err := unstartedProfiler()
+	require.NoError(b, err)
 	b.ReportAllocs()
 	b.ResetTimer()
 
 	for i := 0; i < b.N; i++ {
 		p.doRequest(bat)
 	}
 }
+
+type testServerWaiter func() (http.Header, map[string]string, []string)
+
+func makeTestServer(t *testing.T, statusCode int) (*httptest.Server, *url.URL, testServerWaiter) {
+	wait := make(chan struct{})
+	var header http.Header
+	fields := make(map[string]string)
+	var tags []string
+
+	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		w.WriteHeader(statusCode)
+		header = req.Header
+		_, params, err := mime.ParseMediaType(req.Header.Get("Content-Type"))
+		if err != nil {
+			t.Fatal(err)
+		}
+		mr := multipart.NewReader(req.Body, params["boundary"])
+		defer req.Body.Close()
+		for {
+			p, err := mr.NextPart()
+			if err == io.EOF {
+				break
+			}
+			if err != nil {
+				t.Fatal(err)
+			}
+			slurp, err := ioutil.ReadAll(p)
+			if err != nil {
+				t.Fatal(err)
+			}
+			switch k := p.FormName(); k {
+			case "tags[]":
+				tags = append(tags, string(slurp))
+			default:
+				fields[k] = string(slurp)
+			}
+		}
+		close(wait)
+	}))
+
+	srvURL, err := url.Parse(srv.URL)
+	if err != nil {
+		srv.Close()
+		t.Fatalf("failed to parse server url")
+		return nil, nil, nil
+	}
+
+	waiter := func() (http.Header, map[string]string, []string) {
+		select {
+		case <-wait:
+			// OK
+			return header, fields, tags
+		case <-time.After(time.Second):
+			t.Fatalf("timeout")
+			return nil, nil, nil
+		}
+	}
+
+	return srv, srvURL, waiter
+}
+
+func makeFakeBatch() batch {
```

Comment:
Might be more straightforward
```suggestion
var testBatch = batch{
```